### ISSUE TITLE
ATC '21 | INFaaS: Automated Model-less Inference Serving

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#239]
https://www.usenix.org/system/files/atc21-romero.pdf

best paper award!

**Comment by vycezhong** [comment]
## Motivation

首先解释一个概念**model-variants**. model-variant是一个模型的version，有以下几个方面:
1. model architecture (ResNet50, VGG16)
2. programming framework (TensorFlow, PyTorch, MXNet)
3. model graph optimizer (TensorRT, Neuron, TVM, XLA)
4. hyperparameters (optimizing for batch size of 1, 4, 8, 16)
5. hardware platforms (Haswell or Skylake CPUs, T4 GPUs, FPGA, Inferentia, TPU ...)

对于21个image classification模型，在AWS EC2的可用硬件上 ，可以产生4032种不同的model-variants. 这里面有很大的search space，包括以下几个维度:
1. latency (590ms ~ 11s)
2. cost ($0.096/hr ~ $3.06/hr)
3. accuracy (56.6% ~ 82.5%)

这个search space还是非常大的, 使得developers很难manually针对不同的inference query的要求来选择合适的model variant. 这和ML-training不太一样，优化可以一开始就决定好(statically). 而inference的动态本质属性使得model variant选择不能statially决定. 

作者把这个search space看作一个机会: 
> it allows us to meet the diverse and varying performance, cost and accuracy requirements of applications, in the face of varying load and hardware resource availability, if only we can select and deploy the right model-variant effective for _each_ query. 

现有的inference serving system都忽略了这个opportunity. (surprisingly!)

要想"select and deploy the right model-variant effective for each query"，有这么几个挑战:
1. select the right model-variant: search space很大，如何针对each query选择正确的那个？
2. reduce cost as load varies: 现有的model-scaling方式基本是VM-scaling和replication-based，都是横向拓展. 他们考虑一种新的model-scaling的方式，叫做`model-level vertical scaling`，就是选择一个不同的model variant来应对load变化.  我们当然希望cost越低越好，这就需要在满足query的要求下（SLO, QPS），尽可能减少cost. 文中给了一个不错的例子. 
![image](https://user-images.githubusercontent.com/25879526/131247179-37da6df0-7ddc-41f0-a702-f13f5f0747dc.png)

比如，Table3中，对于第一种情况(QPS=10, SLO=300)，就适合用两个Variant A (QPS = 5, SLO=200). 其他情况类似. 这其实是一个优化问题. 

3. improving utilization at low load: 通过multi-tenancy, share resources. 问题是workload会有干扰，而且随着QPS增加，干扰会越来越明显. 如下图. 
![image](https://user-images.githubusercontent.com/25879526/131247164-c7a8a456-b970-44c6-945e-cc2ee526cf6b.png)

**Comment by vycezhong** [comment]
## Design 

首先介绍了design principles:
1. **Model-less API**: 用户不需要指定模型、模型优化、硬件平台、autoscaling配置. 用户只需要提供high-level的要求，包括performance, cost和accuracy. 
2. INFasS应该自动并且高效地选择一个model-variant，并且能考虑到如何autoscale. 
3. 为了提高资源利用率，应当share hardware resources.
4. 系统应当modular and extensible. 

遵循这些原则，上文提到的挑战都能解决. 


### Model-less API

![image](https://user-images.githubusercontent.com/25879526/131251788-3c7a80e0-a3ea-4198-81c2-dd6012f4b416.png)

看上去对于query有两种API: 
- specify application要求: 需要给定appID, latency和accuracy. 
- specify 一个具体的模型: 这是目前inference serving system所提供的. 给用户direct control. 

### Architecture 

![image](https://user-images.githubusercontent.com/25879526/131251907-c712b39a-0e6e-453a-beb4-1a1849f1489d.png)

- Controller: 包括三个部分
    - Dispatcher: 根据`model-variant selection policy`来选择model
    - VM-Autoscaler: 自动扩容缩容
    - Model Register
- Worker: 执行controller分配的inference queries 
    - Hardware Executor: 管理model-variants的部署和执行. 
    - Dispatcher: 将每一个query通过hardware executor转发到指定的model-variant实例上. 
    - Model-Autoscaler: 检测load变化，决定是否复制running variants或者选择另外一个variant. 也通过`model-variant selection policy`来选择要scale的variant. 
    - Monitoring Daemon: 跟踪机器利用率，管理共享资源，避免违反SLO

**Comment by vycezhong** [comment]
## Selecting and Scaling Model-Variants

这一节讲policy. **区分policy和mechanism非常重要.** 切记. 

INFaaS提供一个internal API, `getVariant`，来调用`model-variant selection policy`. 

注意，policy仅考虑statically-profiled metadata是不够的. 因为以下几个方面(dynamic state of model-variant)也会影响performance和cost:
1. 可能还没load，需要考虑到load latency
2. 可能已经load但是已经满载 (serving at peak throughput)
3. 可能已经load但是受到co-located job的资源竞争
4. 可能由于资源不足的原因没有load

**State machine for the lifecycle of model-variants**
用state machine来捕捉model-variants不断改变的状态. 
![image](https://user-images.githubusercontent.com/25879526/131281440-41a7d892-710d-4772-93a9-f2ac6e7cf9d5.png)

**Comment by vycezhong** [comment]
anyway. 这篇model-less真的很有意思. 用户不需要指定模型，只需要指定high-level的latency和accuracy要求. INFaaS自动去做selection. 妙啊. 多学学best paper思路.

**Comment by vycezhong** [comment]
突然发现这篇19年就挂在arxiv上了，我当时还读过....没想到投了两年才中，一中就是best paper.
